tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,288 @@
+    test_fn(&ServerKey::unchecked_scalar_mul, &|ct_noise, scalar| {
+        ct_noise * scalar as usize
+    });
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
+        test_fn(
+            &ServerKey::unchecked_scalar_left_shift,
+            &|ct_noise, scalar| ct_noise * (1 << scalar as usize),
+        );
+    }
+    test_fn(&ServerKey::unchecked_scalar_right_shift, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+}
+
+fn test_ct_scalar_op_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext, scalar: u8) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &mut Ciphertext, u8),
+                   predicate: &dyn Fn(NoiseLevel, u8) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        op(sk, &mut clone, scalar);
+        assert!(clone.noise_level() == predicate(ct.noise_level(), scalar));
+    };
+
+    test_fn(&ServerKey::unchecked_scalar_add_assign, &|ct_noise, _| {
+        ct_noise
+    });
+    test_fn(&ServerKey::unchecked_scalar_sub_assign, &|ct_noise, _| {
+        ct_noise
+    });
+    test_fn(
+        &ServerKey::unchecked_scalar_mul_assign,
+        &|ct_noise, scalar| ct_noise * scalar as usize,
+    );
+    if scalar != 0 {
+        test_fn(&ServerKey::unchecked_scalar_div_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+        test_fn(&ServerKey::unchecked_scalar_mod_assign, &|_, _| {
+            NoiseLevel::NOMINAL
+        });
+    }
+    test_fn(&ServerKey::unchecked_scalar_bitand_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    test_fn(&ServerKey::unchecked_scalar_bitxor_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+    if scalar < 8 {
+        test_fn(
+            &ServerKey::unchecked_scalar_left_shift_assign,
+            &|ct_noise, scalar| ct_noise * (1 << scalar as usize),
+        );
+    }
+    test_fn(&ServerKey::unchecked_scalar_right_shift_assign, &|_, _| {
+        NoiseLevel::NOMINAL
+    });
+}
+
+#[test]
+fn test_noise_level_propagation() {
+    let keys = KEY_CACHE.get_from_param(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+    let (ck, sk) = (keys.client_key(), keys.server_key());
+
+    let modulus: u64 = 4;
+
+    for _ in 0..10 {
+        let trivial = sk.create_trivial(rand::random::<u64>() % modulus);
+
+        let ct1 = ck.encrypt(rand::random::<u64>() % modulus);
+        let ct2 = sk.add(&ct1, &ct1);
+
+        for ct in [&trivial, &ct1, &ct2] {
```

Comment:
I think here we have that trivial has its noise at 0, ct1 has its noise at nominal, ct2 also has its noise as nominal since its the result from a default ops, which is made using PBS

It would be better if ct2 had non nominal noise (which Is what you meant I think), so it should be `let ct2 = sk.unchecked_add(&ct1, &ct1)`;

Reply:
Indeed, that's what I had in mind
Thanks!